nighthawk/source/client/benchmark_client_impl.cc

```diff
@@ -0,0 +1,226 @@
+      cluster_, uri_->host_and_port(), target_address_,
+      envoy::api::v2::core::Metadata::default_instance(), 1 /* weight */,
+      envoy::api::v2::core::Locality(),
+      envoy::api::v2::endpoint::Endpoint::HealthCheckConfig::default_instance(), 0,
+      envoy::api::v2::core::HealthStatus::HEALTHY)};
+
+  Envoy::Network::ConnectionSocket::OptionsSharedPtr options =
+      std::make_shared<Envoy::Network::ConnectionSocket::Options>();
+  if (use_h2_) {
+    pool_ = std::make_unique<Envoy::Http::Http2::ProdConnPoolImpl>(
+        dispatcher_, host, Envoy::Upstream::ResourcePriority::Default, options);
+  } else {
+    pool_ = std::make_unique<Envoy::Http::Http1::ProdConnPoolImpl>(
+        dispatcher_, host, Envoy::Upstream::ResourcePriority::Default, options);
+  }
+  return true;
+}
+
+void BenchmarkClientHttpImpl::terminate() { pool_.reset(); }
+
+StatisticPtrMap BenchmarkClientHttpImpl::statistics() const {
+  StatisticPtrMap statistics;
+  statistics[connect_statistic_->id()] = connect_statistic_.get();
+  statistics[response_statistic_->id()] = response_statistic_.get();
+  return statistics;
+};
+
+bool BenchmarkClientHttpImpl::tryStartOne(std::function<void()> caller_completion_callback) {
+  ASSERT(target_address_.get());
+  if (!cluster_->resourceManager(Envoy::Upstream::ResourcePriority::Default)
+           .pendingRequests()
+           .canCreate() ||
+      // In closed loop mode we want to be able to control the pacing as exactly as possible.
+      // In open-loop mode we probably want to skip this.
+      // NOTE(oschaaf): We can't consistently rely on resourceManager()::requests() because that
+      // isn't used for h/1 (it is used in tcp and h2 though).
+      ((requests_initiated_ - requests_completed_) >= connection_limit_)) {
+    return false;
+  }
+
+  auto stream_decoder = new StreamDecoder(
+      dispatcher_, api_.timeSource(), *this, std::move(caller_completion_callback),
+      *connect_statistic_, *response_statistic_, request_headers_, measureLatencies());
+  requests_initiated_++;
+  pool_->newStream(*stream_decoder, *stream_decoder);
+  return true;
+}
+
+std::string BenchmarkClientHttpImpl::countersToString(CounterFilter filter) const {
+  auto counters = store_->counters();
+  std::vector<std::string> arr;
+
+  for (auto stat : counters) {
+    if (filter(stat->name(), stat->value())) {
+      arr.push_back(fmt::format("{}:{}", stat->name(), stat->value()));
+    }
+  }
+
+  std::sort(arr.begin(), arr.end());
+  return absl::StrJoin(arr, "\n");
+}
+
+void BenchmarkClientHttpImpl::onComplete(bool success, const Envoy::Http::HeaderMap& headers) {
+  requests_completed_++;
+  if (!success) {
+    // TODO(oschaaf): this information must bubble up!
+    stream_reset_count_++;
+  } else {
+    ASSERT(headers.Status());
+    const int64_t status = Envoy::Http::Utility::getResponseStatus(headers);
+    if (status >= 400 && status <= 599) {
+      // TODO(oschaaf): Figure out why this isn't incremented for us.
```

Comment:
Happy to do some interactive debug on Slack on this one.

Reply:
After looking at the code, it looks like this needs to be changed, the stat that is used is incremented upon actual protocol violations, not when the protocol indicates an error.

Reply:
Added NH's own set of statistics to track counts for 1xx,2xx,..,5xx, resolving this.